### ISSUE TITLE
Proposal: Tracer Components

### ISSUE THREAD

**Issue by discostu105** [open-telemetry/opentelemetry-specification#10]
Proposal: Instead of a global tracer singleton, every library needs to create it's own tracer, which also has a proper name that identifies the library/component that uses the tracer.

Motivation: "Configurability of library/component instrumentation"
 * Identify the library/component that created a span
 * Be able to turn the instrumentation of a certain library/component on/off. This might be necessary if the instrumentation
   * causes too much overhead
   * is buggy (e.g. does not call span.end in every case)
   * avoid "double-tracing". E.g. a tracing vendor (e.g. Dynatrace) might already have built-in auto-instrumentation for a certain library. In this case the vendor might want to decide (automatically or config-based) which of the implementations should be enabled/disabled.

Technically, this could be achieved similar to how logging frameworks deal with Loggers and LoggerFactories ([log4j](https://www.slf4j.org/apidocs/org/slf4j/LoggerFactory.html)).

```
Tracer tracer = OpenConsensus.Tracing.getTracer("io.opensomething.contrib.mongo");
```

Every span would contain the information about the Tracer that it was created with. Based on that a span can be discarded or not.

Not sure how a naming convention could look like. Not sure if "component" is the right word, as it has already been used in the past as attribute name. At Dynatrace we call that a "Sensor". Not sure about granularity of "Tracers". Should every class have one? Should `Tracing` act as a registry (cache already created instances) or as a factory (just create new ones at every call)?

This is kind of related to the GlobalTracer discussion (#107).

**Comment by bogdandrutu** [comment]
@jmacd that is not really a ResourceBuilder but I think offering a simpler API to create Resource (or even a real builder) should be on our list.

**Comment by alramlechner** [comment]
From our (Dynatrace) perspective, the component should not be optional. OTel users should be enforced to provide a component name. 
I would change the sample posted by @tedsuo therefore in to following way:

`tracer = OpenTelemetry.getTracer("io.opensomething.contrib.mongo").toBuilder().
                                      setResource(resource).
                                      build();
`

of course: the whole builder is optional from component from component perspective.

**Comment by discostu105** [comment]
I think there was a bit of a mixup of my intents of this "component" (sorry for late reply). Maybe name is misleading. I actually did not mean to identify *"the component being traced"*, but rather *"the tracing-implementation"*. That's why the existing "component" tag from OpenTracing is not really what I meant here.

So, IMHO this should not be conflated with "resources", as these are in my understanding, rather identifying something like a host/process/service/library.

Knowing the _instrumentation-component_ for every span would allow our agent to determine which instrumentation-components are active in an application at runtime. Based on that we could, at runtime, decide which instrumentation-components we would actually use. This could be configuration based (user opt-in per instrumentation-component) or we could disable specific instrumentations (e.g. faulty/bad instrumentation) based on such component information.

Another use-case is, we could identify certain (well known) instrumentation-components that would overlap with our own already built-in auto-instrumentation and auto-blacklist them (or let the user decide which instrumentation to use) to avoid "double-tracing".

Such concerns have also been brought up recently in the Auto-Instrumentation SIG (see  https://docs.google.com/document/d/1sovSQIGdxXtsauxUNp4qUMEIJZzObdukzPT52eyPCHM/edit#heading=h.dcogdcm13u12)

> A way to discover and avoid potential conflicts/overlap/redundancy between explicit whitebox instrumentation and blackbox instrumentation of the same libraries/packages
>  * That is, if a developer has already added the “official” OpenTelemetry plugin for, say, gRPC, then when the blackbox instrumentation effort adds gRPC support, it should *not* “double-instrument” it and create a mess of extra spans/etc

**Comment by AloisReitbauer** [comment]
Trying to get the Gist of the discussion again. We need to know two additional things about each span that gets generated:

* ```component``` representing the traced component like grpc, jdbc, ...
* ```instrumentation-provider``` this represent the actual instrumentation or contrib library and is needed to turn off instrumentation in case there are problems with it. 

In order to get this information into a span we have three options:
* pass it to a ```Span``` at every call. This is error-prone and chatty
* pass it to the ```span builder```. This would help. However, a ```SpanBuilder``` has the same lifetime as a ```span``` so this would not really help
* pass it to the component that creates a ```SpanBuilder```.  This would mean passing it to the ```Tracer```. this sounds like the most reasonable approach. However, it would not work today because tracers are references to a global tracer object.

We could overcome this problem by creating a ```TracerFactory``` or alternatively a ```TracerBuilder``` as mentioned above. This component would be configured with the required parameters and ensure they are passed to the ```SpanBuilder``` and eventually the ```Span```. 

Does this capture the problem and proposed solutions properly?

**Comment by tedsuo** [comment]
@AloisReitbauer this does capture the issue well. Modeling a `service` as a set of resources, and then each `component` as a nested set of  resources feels correct. 

The only complication is how the Tracer object relates context propagation. But both traces and metrics will want to be labelled with this information. So there's still some work to be done, in general, to cleanly separate the context propagation layer from the observability layer.

**Comment by lizthegrey** [comment]
+1 that this would be extremely useful to us over at Honeycomb - cc @maplebed

We sometimes like to write to more than one distinct dataset or even telemetry processing environment from some of our internal tools.

**Comment by discostu105** [comment]
@z1c0 wrote an RFC that covers this issue: https://github.com/open-telemetry/rfcs/pull/16

**Comment by Oberon00** [comment]
I think this can be closed once #264 is merged.

**Comment by tedsuo** [comment]
Closing this now. Thank you all, for taking this over the finish line.
